qos/evm/evm.go

```diff
@@ -1,136 +1,32 @@
-
-	body, err := io.ReadAll(req.Body)
-	if err != nil {
-		return relayer.Payload{}, err
-	}
-
-	if p.requestTimeout == 0 {
-		p.requestTimeout = defaultEVMRequestTimeout
-	}
-
-	return relayer.Payload{
-		Data:            string(body),
-		Method:          req.Method,
-		Path:            req.URL.Path,
-		TimeoutMillisec: int(p.requestTimeout.Milliseconds()),
-	}, nil
-}
-
-/* EVM QoS Response Builder */
-
-type EVMResponseBuilder struct{}
-
-func (b *EVMResponseBuilder) GetHTTPResponse(ctx context.Context, resp relayer.Response) (gateway.HTTPResponse, error) {
-	return &EVMHTTPResponse{
-		payload:        resp.Bytes,
-		httpStatusCode: resp.HTTPStatusCode,
-		httpHeaders:    http.Header{},
-	}, nil
-}
-
-// TODO_INCOMPLETE: This method needs to validate the response is a valid JSON-RPC object
-func (b *EVMResponseBuilder) GetHTTPErrorResponse(ctx context.Context, err error) gateway.HTTPResponse {
-
-	return &EVMHTTPResponse{
-		payload: []byte(fmt.Sprintf(evmErrorTemplate, err.Error())),
-		// TODO_TECHDEBT: get HTTP status code from specific error type
-		httpStatusCode: http.StatusInternalServerError,
-		httpHeaders:    http.Header{},
-	}
-}
-
-/* EVM Endpoint Selector */
-
-// TODO_INCOMPLETE: implement a proper endpoint selector for EVM chains
-type EVMEndpointSelector struct{}
-
-// TODO_INCOMPLETE: implement actual logic for selecting the most appropriate endpoint using QoS metrics
-func (s *EVMEndpointSelector) Select(endpoints map[relayer.AppAddr][]relayer.Endpoint) (relayer.AppAddr, relayer.EndpointAddr, error) {
-	for appAddr, endpointList := range endpoints {
-		if len(endpointList) > 0 {
-			randomIndex := rand.Intn(len(endpointList))
-			return appAddr, endpointList[randomIndex].Addr(), nil
-		}
-	}
-	return "", "", fmt.Errorf("no endpoints available")
-}
-
-/* EVM HTTP Response */
-
-type EVMHTTPResponse struct {
-	payload        []byte
-	httpStatusCode int
-	httpHeaders    http.Header
-}
-
-func (r *EVMHTTPResponse) GetPayload() []byte {
-	return r.payload
-}
-
-func (r *EVMHTTPResponse) GetHTTPStatusCode() int {
-	return r.httpStatusCode
-}
-
-func (r *EVMHTTPResponse) GetHTTPHeaders() map[string]string {
-	headersMap := make(map[string]string)
-	for key, value := range r.httpHeaders {
-		headersMap[key] = value[0]
+// QoS struct performs the functionality defined by
+// gateway package's ServiceQoS, which consists of:
+// A) a QoSRequestParser which builds EVM-specific
+// ServiceRequestContext objects by parsing user HTTP
+// requests.
+// B) an EndpointSelector, which selects an endpoint
+// for performing a service request.
+var _ gateway.QoSService = &QoS{}
+
+// QoS is the ServiceQoS implementations for EVM-based chains.
+// It contains logic specific to EVM-based chains,
+// including request parsing, response building,
+// and endpoint validation/selection.
+type QoS struct {
+	endpointStore *EndpointStore
+	logger        polylog.Logger
+}
+
+func NewServiceQoS(endpointStore *EndpointStore, logger polylog.Logger) *QoS {
```

Comment:
Why are we returning the struct rather than the interface?

Reply:
It is intentional, and matches the general pattern used on PATH: producers build and return structs, consumers define and accept interfaces.